File: interceptor/crypto.c
#include "crypto.h"
#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static uint8_t link_key[16];
static int key_initialized = 0;

int crypto_init_link_key(const char *key_hex) {
    if (strlen(key_hex) != 32) {
        ERROR_PRINT("Invalid link key length (expected 32 hex chars)");
        return -1;
    }
    
    for (int i = 0; i < 16; i++) {
        char byte_str[3] = {key_hex[i*2], key_hex[i*2+1], '\0'};
        link_key[i] = (uint8_t)strtol(byte_str, NULL, 16);
    }
    
    key_initialized = 1;
    INFO_PRINT("Link key initialized");
    return 0;
}

int crypto_decrypt_payload(const uint8_t *encrypted, size_t len, uint8_t *decrypted) {
    if (!key_initialized) {
        ERROR_PRINT("Link key not initialized");
        return -1;
    }
    
    DEBUG_PRINT("Decrypting %zu bytes (E0 cipher not fully implemented - copying data)", len);
    memcpy(decrypted, encrypted, len);
    
    return len;
}

int crypto_encrypt_payload(const uint8_t *plain, size_t len, uint8_t *encrypted) {
    if (!key_initialized) {
        ERROR_PRINT("Link key not initialized");
        return -1;
    }
    
    DEBUG_PRINT("Encrypting %zu bytes (E0 cipher not fully implemented - copying data)", len);
    memcpy(encrypted, plain, len);
    
    return len;
}

File: btintercept/src/crypto.rs
use std::sync::atomic::{AtomicBool, Ordering};

use crate::{debug_print, error_print, info_print};

static KEY_INITIALIZED: AtomicBool = AtomicBool::new(false);
static mut LINK_KEY: [u8; 16] = [0u8; 16];

pub fn init_link_key(key_hex: &str) -> Result<(), String> {
    if key_hex.len() != 32 {
        error_print!("Invalid link key length (expected 32 hex chars)");
        return Err("Invalid link key length".to_string());
    }

    unsafe {
        for i in 0..16 {
            let byte_str = &key_hex[i * 2..i * 2 + 2];
            LINK_KEY[i] = u8::from_str_radix(byte_str, 16)
                .map_err(|e| format!("Invalid hex char: {}", e))?;
        }
    }

    KEY_INITIALIZED.store(true, Ordering::SeqCst);
    info_print!("Link key initialized");
    Ok(())
}

// E0 cipher not fully implemented — passes data through unchanged (same as C stub).
pub fn decrypt_payload(encrypted: &[u8]) -> Result<Vec<u8>, String> {
    if !KEY_INITIALIZED.load(Ordering::SeqCst) {
        error_print!("Link key not initialized");
        return Err("Link key not initialized".to_string());
    }
    debug_print!(
        "Decrypting {} bytes (E0 cipher not fully implemented - copying data)",
        encrypted.len()
    );
    Ok(encrypted.to_vec())
}

pub fn encrypt_payload(plain: &[u8]) -> Result<Vec<u8>, String> {
    if !KEY_INITIALIZED.load(Ordering::SeqCst) {
        error_print!("Link key not initialized");
        return Err("Link key not initialized".to_string());
    }
    debug_print!(
        "Encrypting {} bytes (E0 cipher not fully implemented - copying data)",
        plain.len()
    );
    Ok(plain.to_vec())
}
